From 9b6528198a88114155be0996f0bbebef165a0ef6 Mon Sep 17 00:00:00 2001
From: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
Date: Fri, 21 Apr 2023 10:08:44 +0800
Subject: [PATCH 4/6] preemption: add lldp verify parameter

The preemption needs lldp verify process according to IEEE 802.3-2018,
This patch add lldp verify parameter on ethtool preemption command.

Signed-off-by: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
---
 ethtool.c                    | 13 +++++++++++--
 netlink/preempt.c            | 12 ++++++++++++
 uapi/linux/ethtool.h         |  3 +++
 uapi/linux/ethtool_netlink.h |  4 +++-
 4 files changed, 29 insertions(+), 3 deletions(-)

diff --git a/ethtool.c b/ethtool.c
index a9f0167..80802ce 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -1648,11 +1648,17 @@ static void dump_fpcmd(struct ethtool_fp *fpcmd)
 		fprintf(stdout, "not supported\n");
 
 	fprintf(stdout, "	status: ");
-	if (fpcmd->fp_enabled)
+	if (fpcmd->fp_status)
 		fprintf(stdout, "enabled\n");
 	else
 		fprintf(stdout, "disabled\n");
 
+	fprintf(stdout, "	active: ");
+	if (fpcmd->fp_active)
+		fprintf(stdout, "active\n");
+	else
+		fprintf(stdout, "not active\n");
+
 	fprintf(stdout, "	supported queues: %#x\n",
 		fpcmd->supported_queues_mask);
 	fprintf(stdout, "	preemptible queues: %#x\n",
@@ -5234,9 +5240,11 @@ static int do_get_preempt(struct cmd_context *ctx)
 static int do_set_preempt(struct cmd_context *ctx)
 {
 	int fp_c = -1, preempt_queues_mask_c = -1, min_frag_c = -1;
+	int lldp_v = -1;
 	int change = -1;
 	struct ethtool_fp fpcmd;
 	struct cmdline_info cmdline_fp[] = {
+		{ "lldp-verify", CMDL_BOOL, &lldp_v, &fpcmd.fp_supported },
 		{ "fp", CMDL_BOOL, &fp_c, &fpcmd.fp_enabled },
 		{ "preemptible-queues-mask", CMDL_U32, &preempt_queues_mask_c,
 		  &fpcmd.preemptible_queues_mask },
@@ -6121,7 +6129,8 @@ static const struct option args[] = {
 		.func	= do_set_preempt,
 		.nlfunc	= nl_set_preempt,
 		.help	= "Set Frame Preemption settings",
-		.xhelp	= "		[ fp on|off ]\n"
+		.xhelp	= "		[ lldp-verify on|off ]\n"
+			  "		[ fp on|off ]\n"
 			  "		[ preemptible-queues-mask %x ]\n"
 			  "		[ min-frag-size %d ]\n",
 	},
diff --git a/netlink/preempt.c b/netlink/preempt.c
index 6f66e43..ee99bb1 100644
--- a/netlink/preempt.c
+++ b/netlink/preempt.c
@@ -56,6 +56,11 @@ int preempt_get_reply_cb(const struct nlmsghdr *nlhdr, void *data)
 		printf("\tsupport: %s\n",
 		       supported ? "supported" : "not supported");
 	}
+	if (tb[ETHTOOL_A_PREEMPT_STATUS]) {
+		int status = mnl_attr_get_u8(tb[ETHTOOL_A_PREEMPT_STATUS]);
+
+		printf("\tstatus: %s\n", status ? "enable" : "not enable");
+	}
 	if (tb[ETHTOOL_A_PREEMPT_ACTIVE]) {
 		int active = mnl_attr_get_u8(tb[ETHTOOL_A_PREEMPT_ACTIVE]);
 
@@ -103,6 +108,13 @@ static const struct lookup_entry_u8 fp_values[] = {
 };
 
 static const struct param_parser set_preempt_params[] = {
+	{
+		.arg		= "lldp-verify",
+		.group		= ETHTOOL_MSG_PREEMPT_SET,
+		.type		= ETHTOOL_A_PREEMPT_LLDP_VERIFY,
+		.handler	= nl_parse_u8bool,
+		.min_argc	= 1,
+	},
 	{
 		.arg		= "fp",
 		.group		= ETHTOOL_MSG_PREEMPT_SET,
diff --git a/uapi/linux/ethtool.h b/uapi/linux/ethtool.h
index 3bbeb26..55ef5ce 100644
--- a/uapi/linux/ethtool.h
+++ b/uapi/linux/ethtool.h
@@ -394,7 +394,10 @@ struct ethtool_eee {
 struct ethtool_fp {
 	__u32	cmd;
 	__u8	fp_supported;
+	__u8	fp_lldp_verify;
 	__u8	fp_enabled;
+	__u8	fp_status;
+	__u8	fp_active;
 	__u32	supported_queues_mask;
 	__u32	preemptible_queues_mask;
 	__u32	min_frag_size;
diff --git a/uapi/linux/ethtool_netlink.h b/uapi/linux/ethtool_netlink.h
index 02a4031..e519843 100644
--- a/uapi/linux/ethtool_netlink.h
+++ b/uapi/linux/ethtool_netlink.h
@@ -572,7 +572,9 @@ enum {
 	ETHTOOL_A_PREEMPT_UNSPEC,		/* nest - _A_HEADER_* */
 	ETHTOOL_A_PREEMPT_HEADER,		/* u8 */
 	ETHTOOL_A_PREEMPT_SUPPORTED,		/* u8 */
-	ETHTOOL_A_PREEMPT_ACTIVE,		/* u32 */
+	ETHTOOL_A_PREEMPT_LLDP_VERIFY,		/* u8 */
+	ETHTOOL_A_PREEMPT_STATUS,		/* u8 */
+	ETHTOOL_A_PREEMPT_ACTIVE,		/* u8 */
 	ETHTOOL_A_PREEMPT_MIN_FRAG_SIZE,	/* u32 */
 	ETHTOOL_A_PREEMPT_QUEUES_SUPPORTED,	/* u32 */
 	ETHTOOL_A_PREEMPT_QUEUES_PREEMPTIBLE,	/* u32 */
-- 
2.25.1

